test/e2e/external_gateways.go

```diff
@@ -0,0 +1,482 @@
+			"annotate",
+			"namespace",
+			f.Namespace.Name,
+			fmt.Sprintf("k8s.ovn.org/routing-external-gws=%s,%s", addresses.gatewayIPs[0], addresses.gatewayIPs[1]),
+		}
+		framework.Logf("Annotating the external gateway test namespace to container gateways: %s, %s", addresses.gatewayIPs[0], addresses.gatewayIPs[1])
+		framework.RunKubectlOrDie(f.Namespace.Name, annotateArgs...)
+
+		expectedHostNames := make(map[string]struct{})
+		for _, c := range []string{gwContainer1, gwContainer2} {
+			res, err := runCommand("docker", "exec", c, "hostname")
+			framework.ExpectNoError(err, "failed to run hostname on", c)
+			hostname := strings.TrimSuffix(res, "\n")
+			framework.Logf("Hostname for %s is %s", c, hostname)
+			expectedHostNames[hostname] = struct{}{}
+		}
+		framework.Logf("Expected hostnames are %v", expectedHostNames)
+
+		returnedHostNames := make(map[string]struct{})
+		success := false
+
+		// Picking only the first address, the one the udp listener is set for
+		target := addresses.targetIPs[0]
+		for i := 0; i < 20; i++ {
+			args := []string{"exec", srcPodName, "--"}
+			if protocol == "tcp" {
+				args = append(args, "bash", "-c", fmt.Sprintf("echo | nc -w 1 %s %d", target, destPort))
+			} else {
+				args = append(args, "bash", "-c", fmt.Sprintf("echo | nc -w 1 -u %s %d", target, destPort))
+			}
+			res, err := framework.RunKubectl(f.Namespace.Name, args...)
+			framework.ExpectNoError(err, "failed to reach ", target, protocol)
+			hostname := strings.TrimSuffix(res, "\n")
+			if hostname != "" {
+				returnedHostNames[hostname] = struct{}{}
+			}
+
+			if cmp.Equal(returnedHostNames, expectedHostNames) {
+				success = true
+				break
+			}
+		}
+
+		framework.Logf("Received hostnames for protocol %s are %v ", protocol, returnedHostNames)
+
+		if !success {
+			framework.Failf("Failed to hit all the external gateways via for protocol %s, diff %s", protocol, cmp.Diff(expectedHostNames, returnedHostNames))
+		}
+
+	}, ginkgotable.Entry("IPV4 udp", &addressesv4, "udp", externalUDPPort),
+		ginkgotable.Entry("IPV4 tcp", &addressesv4, "tcp", externalTCPPort),
+		ginkgotable.Entry("IPV6 udp", &addressesv6, "udp", externalUDPPort),
+		ginkgotable.Entry("IPV6 tcp", &addressesv6, "tcp", externalTCPPort))
+})
+
+// setupGatewayContainers sets up external containers, adds routes to the nodes, sets up listeners.
+// All its needed for namespace / pod gateway tests.
+func setupGatewayContainers(f *framework.Framework, gwContainer1, gwContainer2, srcPodName string, updPort, tcpPort, numOfIPs int) (gatewayTestIPs, gatewayTestIPs) {
+	addressesv4 := gatewayTestIPs{targetIPs: make([]string, 0)}
+	addressesv6 := gatewayTestIPs{targetIPs: make([]string, 0)}
+	// retrieve worker node names
+	nodes, err := e2enode.GetBoundedReadySchedulableNodes(f.ClientSet, 3)
+	framework.ExpectNoError(err)
+	if len(nodes.Items) < 3 {
+		framework.Failf(
+			"Test requires >= 3 Ready nodes, but there are only %v nodes",
+			len(nodes.Items))
+	}
```

Comment:
some places have this code in the BeforeEach block too, we should decide in one place only and make it consistent. I think that it fits better in the BeforeEach block, `setupGatewayContainer` should asume the environment is ready

Reply:
Will do.